pkg/globalnet/controllers/node_controller.go

```diff
@@ -0,0 +1,220 @@
+	var err error
+
+	klog.Info("Creating Node controller")
+
+	iptIface, err := iptables.New()
+	if err != nil {
+		return nil, errors.WithMessage(err, "error creating the IPTablesInterface handler")
+	}
+
+	controller := &nodeController{
+		baseIPAllocationController: newBaseIPAllocationController(pool, iptIface),
+	}
+
+	controller.nodeName = nodeName
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Node syncer",
+		ResourceType:        &corev1.Node{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     corev1.NamespaceAll,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewUpdateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		ResourcesEquivalent: controller.onNodeUpdated,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, gvr, err := admUtil.ToUnstructuredResource(&corev1.Node{}, config.RestMapper)
+	if err != nil {
+		return nil, err
+	}
+
+	federator := federate.NewUpdateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll)
+	controller.nodes = config.SourceClient.Resource(*gvr)
+	localNodeInfo, err := controller.nodes.Get(context.TODO(), controller.nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.WithMessagef(err, "error retrieving localNodeInfo for %q", controller.nodeName)
+	}
+
+	if localNodeInfo != nil {
+		if err := controller.reserveAllocatedIPs(federator, localNodeInfo); err != nil {
+			return nil, err
+		}
+	}
+
+	return controller, nil
+}
+
+func (n *nodeController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	node := from.(*corev1.Node)
+
+	// If the event corresponds to a different node which has globalIP annotation, release the globalIP back to Pool.
+	if node.Name != n.nodeName {
+		if existingGlobalIP := node.GetAnnotations()[SmGlobalIP]; existingGlobalIP != "" {
+			if op == syncer.Delete {
+				_ = n.pool.Release(existingGlobalIP)
+				return nil, false
+			} else {
+				_ = n.pool.Release(existingGlobalIP)
+				return n.updateNodeAnnotation(node, "")
+			}
+		}
+
+		return nil, false
+	}
+
+	cniIfaceIP := node.GetAnnotations()[constants.CNIInterfaceIP]
+	if cniIfaceIP == "" {
+		// To support connectivity from HostNetwork to remoteCluster, globalnet requires the
+		// cniIfaceIP of the respective node. Route-agent running on the node annotates the
+		// respective node with the cniIfaceIP. In this API, we check for the presence of this
+		// annotation and process the node event only when the annotation exists.
+		return nil, false
+	}
+
+	return n.onCreateOrUpdate(node)
+}
+
+func (n *nodeController) onCreateOrUpdate(node *corev1.Node) (runtime.Object, bool) {
+	klog.Infof("Node %q is updated", node.Name)
```

Comment:
This could be a create - how about we move this above L116?

`klog.Infof("Processing %sd Node %q", op, node.Name)`